R/test_nse.R

```diff
@@ -0,0 +1,45 @@
+#' @export
+data_addsuffix_test <- function(data, pattern, select = NULL, exclude = NULL, ignore_case = FALSE, ...) {
+
+  select <- test_fun(select, data, exclude, ignore_case)
+
+  selected_columns <- colnames(data) %in% select
+  colnames(data)[selected_columns] <- paste0(colnames(data)[selected_columns], pattern)
+  data
+}
+
+
+test_fun <- function(select, data, exclude, ignore_case) {
+  fixed <- TRUE
+  # avoid conflicts
+  conflicting_packages <- .conflicting_packages("poorman")
+
+  # in case pattern is a variable from another function call...
+  p <- try(eval(select), silent = TRUE)
+  if (inherits(p, c("try-error", "simpleError"))) {
+    p <- substitute(select)
+  }
+
+  # check if pattern is a function like "starts_with()"
+  select <- tryCatch(eval(p), error = function(e) NULL)
+
+  # if select could not be evaluated (because expression "makes no sense")
+  # try to evaluate and find select-helpers. In this case, set fixed = FALSE,
+  # so we can use grepl()
+  if (is.null(select)) {
+    evaluated_pattern <- .evaluate_pattern(insight::safe_deparse(p), data, ignore_case = ignore_case)
+    select <- evaluated_pattern$pattern
+    fixed <- evaluated_pattern$fixed
+  }
+
+  # seems to be no valid column name or index, so try to grep
+  if (isFALSE(fixed)) {
+    select <- colnames(data)[grepl(select, colnames(data), ignore.case = ignore_case)]
+  }
+
+  # load again
+  .attach_packages(conflicting_packages)
```

Comment:
This is a problem is e.g., `poorman` was loaded before `dplyr`, so for the used before running this function dplyr masks poorman, but after running it, poorman masks dplyr...

Reply:
That's unrelated. The problem is that NSE is not working when passing down `select` to a helper function:

``` r
library(datawizard)
data_addsuffix(iris, "HA", col_starts_with("Sepal")) |> head()
#>   Sepal.LengthHA Sepal.WidthHA Petal.Length Petal.Width Species
#> 1            5.1           3.5          1.4         0.2  setosa
#> 2            4.9           3.0          1.4         0.2  setosa
#> 3            4.7           3.2          1.3         0.2  setosa
#> 4            4.6           3.1          1.5         0.2  setosa
#> 5            5.0           3.6          1.4         0.2  setosa
#> 6            5.4           3.9          1.7         0.4  setosa
data_addsuffix_test(iris, "HA", col_starts_with("Sepal")) |> head()
#> Warning in eval(p): restarting interrupted promise evaluation

#> Warning in eval(p): restarting interrupted promise evaluation
#>   Sepal.Length Sepal.Width Petal.Length Petal.Width Species
#> 1          5.1         3.5          1.4         0.2  setosa
#> 2          4.9         3.0          1.4         0.2  setosa
#> 3          4.7         3.2          1.3         0.2  setosa
#> 4          4.6         3.1          1.5         0.2  setosa
#> 5          5.0         3.6          1.4         0.2  setosa
#> 6          5.4         3.9          1.7         0.4  setosa
```

<sup>Created on 2022-03-18 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>